src/borg/key.py

```diff
@@ -147,6 +150,75 @@ def decrypt(self, id, data, decompress=True):
+        except InvalidRPCMethod as error:
+            # old server version, suppress further calls
+            self.get_repo_free_nonce = lambda: None
+            self.commit_repo_nonce_reservation = lambda next_unreserved, start_nonce: None
+            return None
+
+    def commit_repo_nonce_reservation(self, next_unreserved, start_nonce):
+        self.repository.commit_nonce_reservation(next_unreserved, start_nonce)
+
+    def ensure_reservation(self, nonce_space_needed):
+        if not self.end_of_nonce_reservation:
+            repo_free_nonce = self.get_repo_free_nonce()
+            local_free_nonce = self.get_local_free_nonce()
+            next_nonce = max(x for x in (repo_free_nonce, local_free_nonce, self.manifest_nonce) if x is not None)
+            reservation_end = next_nonce + nonce_space_needed + NONCE_SPACE_RESERVATION
+            assert reservation_end < MAX_REPRESENTABLE_NONCE
+            self.commit_repo_nonce_reservation(reservation_end, repo_free_nonce)
+            self.commit_local_nonce_reservation(reservation_end, local_free_nonce)
+            self.end_of_nonce_reservation = reservation_end
+            self.enc_cipher.reset(None, next_nonce.to_bytes(16, byteorder='big'))
+        else:
+            next_nonce = int.from_bytes(self.enc_cipher.iv, byteorder='big')
+            assert next_nonce <= self.end_of_nonce_reservation
+            if next_nonce + nonce_space_needed > self.end_of_nonce_reservation:
+                # try to expand reservation
+                repo_free_nonce = self.get_repo_free_nonce()
+                local_free_nonce = self.get_local_free_nonce()
+                free_nonce_space = max(x for x in (repo_free_nonce, local_free_nonce, self.manifest_nonce, self.end_of_nonce_reservation) if x is not None)
+                if free_nonce_space == self.end_of_nonce_reservation:
+                    # nothing else got an reservation, just extend
+                    reservation_end = self.end_of_nonce_reservation + nonce_space_needed + NONCE_SPACE_RESERVATION
+                    assert reservation_end < MAX_REPRESENTABLE_NONCE
+                    self.commit_repo_nonce_reservation(reservation_end, repo_free_nonce)
+                    self.commit_local_nonce_reservation(reservation_end, local_free_nonce)
+                    self.end_of_nonce_reservation = reservation_end
+                else:
+                    # need an entirly new reservation
+                    reservation_end = free_nonce_space + nonce_space_needed + NONCE_SPACE_RESERVATION
+                    assert reservation_end < MAX_REPRESENTABLE_NONCE
+                    self.commit_repo_nonce_reservation(reservation_end, repo_free_nonce)
+                    self.commit_local_nonce_reservation(reservation_end, local_free_nonce)
+                    self.end_of_nonce_reservation = reservation_end
+                    self.enc_cipher.reset(None, free_nonce_space.to_bytes(16, byteorder='big'))
```

Comment:
actually if the comments about init and expanding and interleaved are considered to be unimportant it could even be for the whole conditional following `if self.end_of_nonce_reservation is None`
```python
        if free_nonce_space != self.end_of_nonce_reservation:
            self.enc_cipher.reset(None, free_nonce_space.to_bytes(16, byteorder='big'))
```
But i think this it would quite a bit less meaningful. Maybe a big comment could help there, but...